GH-603: Initial workspace API changes to support unset root case.

kittaakos opened the pull request:
Task: #603.
Closes: #603.

akosyakov commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -16,48 +16,79 @@ import { WorkspaceServer } from "../common";
+                    if (uri) {
+                        self.validateRoot(uri).then(
+                            root => resolve(root),
+                            () => setTimeout(waitUntilResolved, 200)
```
Would it be possible to use JSON-RPC notification instead of polling?

kittaakos replied:
If you can provide an example, sure, I will change. Thanks!

akosyakov replied:
I mean to add `WorkspaceClient` similar to `ILoggerClient`: 
```ts
    constructor(
        @inject(FileSystem) protected readonly fileSystem: FileSystem,
        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
        @inject(WorkspaceServer) protected readonly server: WorkspaceServer
    ) {
        this.root = server.getRoot().then(current =>
            current ? this.validateRoot(current) : new Promise(resolve => {
                server.setClient({
                    onDidChangeRoot: uri => {
                        if (uri) {
                            resolve(this.validateRoot(uri));
                        }
                    }
                });
            )
        );
    }
```

kittaakos replied:
Thanks for the snippet! Well, yeah. Could be. Exactly this is what I wanted to avoid, boilerplate client, proxy, server, binding, blabla... just for listening to a single property change.

akosyakov replied:
Should we start using polling when we have 2 directional connection?

kittaakos replied:
> Would it be possible to

:) Sure, then in the future please write: _Replace it with JSON-RPC notification... Here comes an example snippet. `// snippet...` Let me know if it needs clarification._

And we can save some time and energy. Thanks!

akosyakov commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -16,48 +16,79 @@ import { WorkspaceServer } from "../common";
+     * Returns a promise resolving to the default workspace root location. This method should be used if [resolvedRoot](WorkspaceService.rootResolved) is `false`.
+     */
+    get defaultRoot(): Promise<FileStat> {
+        return new Promise((resolve, reject) => {
```
an outer promise is redundant?

hexa00 commented on packages/workspace/src/node/default-workspace-server.ts:
```diff
@@ -46,32 +46,31 @@ export class WorkspaceCliContribution implements CliContribution {
+        this.root = this.retrieveRootURI();
+    }
+
+    getRoot(): Promise<string | undefined> {
```
Use a getter here?

akosyakov replied:
I don't think that getters are supported for JSON-RPC implementation

hexa00 replied:
ha right, forgot we were in a server for a second :)

hexa00 commented on packages/workspace/src/node/default-workspace-server.ts:
```diff
@@ -46,32 +46,31 @@ export class WorkspaceCliContribution implements CliContribution {
-    }
-
-    protected async getRootURI(): Promise<string> {
+    protected async retrieveRootURI(): Promise<string | undefined> {
```
This is a bit weird retrieve from what?.. 
I would suggest `getRootFromCLI` ?

hexa00 commented on packages/workspace/src/node/default-workspace-server.ts:
```diff
@@ -46,32 +46,31 @@ export class WorkspaceCliContribution implements CliContribution {
         @inject(ILogger) protected readonly logger: ILogger
     ) {
-        this.root = this.getRootURI();
+        this.root = this.retrieveRootURI();
```
I find it really hard to tell if something is a uri as a string when it's not encoded in the type or in the value name.
I guess this is the reason we have functions called `getRootURI` or `retrieveRootURI`
Feels like the value should be thus called `rootURI` ?

hexa00 commented on packages/workspace/src/common/workspace-protocol.ts:
```diff
@@ -12,12 +12,23 @@ export const workspacePath = '/services/workspace';
+    /**
+     * Returns with a promise that resolves to the workspace root URI as a string. Resolves to `undefined` if the workspace root is not yet set.
+     */
+    getRoot(): Promise<string | undefined>;
```
as per comment below ... consider getRootURI

akosyakov replied:
not sure about it, for protocols URI is always `string`

hexa00 replied:
Yes the issue is more what is root ? is a URI or an fs path ? You have to open the code / jsdoc to figure it out.

akosyakov replied:
I don't think we send paths over JSON-RPC, only normalized URIs.

hexa00 commented on packages/workspace/src/common/workspace-protocol.ts:
```diff
@@ -12,12 +12,23 @@ export const workspacePath = '/services/workspace';
+     * underlying process or any other arbitrary logic.
      */
-    getRoot(): Promise<string>;
+    getDefaultRoot(): Promise<string>;
```
Ditto URI

akosyakov commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -35,19 +36,24 @@ export class TerminalFrontendContribution implements CommandContribution, MenuCo
-            execute: () => this.newTerminal()
+        this.workspaceService.root.then(() => {
+            commands.registerCommand(TerminalCommands.NEW, {
+                isEnabled: () => true,
```
Should we have a single provider of default workspace root for all clients? Instead of each client fallbacks to its default? e.g. if we figure out that for electron `electron.process.cwd` should be used as a default, only one place should be changed.

hexa00 replied:
I'm not sure, I actually thought the idea of the PR was to have undefined as default pretty much.
I don't think it makes sense to open cwd as a default workspace, especially given that we inotify this root dir... if the user opens it from $HOME or /usr/bin... this will be bad.

It does make sense for the terminal however. 

In general I don't think the terminal should be linked to the workspace except to provide the convenience of starting in the root dir if there is one.

akosyakov replied:
> It does make sense for the terminal to use cwd however.

It won't work nicely for the packaged electron. VS code opens the terminal in the user home dir if a folder is not opened and otherwise in the workspace root.

> In general I don't think the terminal should be linked to the workspace except to provide the convenience of starting in the root dir if there is one.

I thought an idea was that:
- components which can work without an opened project should use `this.tryRoot` and fall back to the common non-project root;
- components, which cannot, should use `this.root` and wait when a project is opened

@kittaakos correct me if I am wrong

hexa00 replied:
@akosyakov  OK this makes more sense to me, thx!

svenefftinge replied:
> It does make sense for the terminal to use cwd however.

Why? You get a terminal to the server's cwd. I don't see why that is better than a workspace root there?

hexa00 replied:
I mean that it's possible to use cwd if you don't have a workspace root. Of course it's better to use the workspace root.

Note that I linke @akosyakov idea.. with the tryRoot I think that's nice.

akosyakov commented on packages/workspace/src/browser/workspace-frontend-contribution.ts:
```diff
@@ -47,18 +47,21 @@ export class WorkspaceFrontendContribution implements CommandContribution, MenuC
-                const node = DirNode.createRoot(startWith);
+        this.workspaceService.rootResolved.then(resolved => {
+            const root = resolved ? this.workspaceService.root : this.workspaceService.defaultRoot;
+            root.then(root => {
```
thought the same, it would be nice to have 2 methods only like `this.root` and `this.tryRoot`, the former is resolved only when the root is set, the latter fallbacks to the default if it is not set

hexa00 replied:
Is there a situation where we would not want the default to be used?

akosyakov replied:
for example for the navigator, I think it should not show up till the real root is set, i.e. some project is opened

kittaakos replied:
> Is there a situation where we would not want the default to be used?

Yes, everywhere. If you do not have a workspace root, then you cannot do anything, right?

The default is needed to feed the workspace root selector dialog.

hexa00 replied:
I guess I'm confused from the PR comment : `Workspace root could be unset (undefined) so downstream services
should handle it gracefully.`

So I thought the idea is that you would not need a default ?

hexa00 replied:
Should not the root selector be independent of that ? and use it's own default like the home directory?

hexa00 replied:
I just read @akosyakov comment about tryRoot and I think I get the idea now... so you can disregard my last comments.

kittaakos commented:
I have updated the [requirements of this task](https://github.com/theia-ide/theia/issues/603#issue-263056123). I have tried to sum up what we've initially planned to achieve to be able to proceed on the electron packaging task. I have also added things that should be implemented/fixed as part of a follow-up task but still related to this topic. 

@svenefftinge, could you please check them too? I might have left out something important from my notes.

If you have any concerns about the goals and approaches, please add your comments to the [ticket](#603). The code related things should still go here. Thanks!

(CC: @hexa00, @akosyakov)

svenefftinge commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -16,48 +16,108 @@ import { WorkspaceServer } from "../common";
+            if (validRoot) {
+                resolve(this.toFileStat(root!));
+            } else {
+                this.rootWasResolvedAtStartup = Promise.resolve(false);
```
reassigning this field will not notify any listeners on the previously assigned promise. I think a Deferred Promise would simplify this code a bit. See `promise-util.ts` in core.

kittaakos commented:
This PR is ready for review. Still waiting for a green build, though.

akosyakov commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -30,8 +30,14 @@ export class FileNavigatorContribution implements FrontendApplicationContributio
         );
-        this.workspaceService.root.then(fileStat => {
-            this.fileNavigator.model.root = DirNode.createRoot(fileStat);
+        this.workspaceService.rootResolved.then(resolved => {
```
I really think adding `tryRoot` (`root` which can reject) instead of `rootResolved` is better API-wise, e.g. this code will be:
```ts
this.workspaceService.tryRoot().then(fileStat => 
    this.fileNavigator.model.root = DirNode.createRoot(fileStat)
).catch(() => 
    this.fileNavigator.update()
);
```
simpler contract, less to write and read

kittaakos replied:
All right, I am opened for a small modification; how should `WorkspaceRootAwareCommandHandler` handle when there is not root? Same applies to the terminal contributions too, for instance. The navigator example is clear.

akosyakov replied:
They continue using `this.root` as is.

I only meant that instead of checking `this.rootResolved` each time before `this.root`, it could be done once in `this.tryRoot` that returns `undefined` (or reject) if the root is not resolved. Returning `undefined` seems to be better looking at https://github.com/theia-ide/theia/pull/615/files#diff-416c9dea7bd5aec1cc7743b90b2c5528R51

kittaakos replied:
I see I got it. Let me try one more iteration on this.

akosyakov replied:
it would be:
```ts
        this.workspaceService.tryRoot().then(async resolvedRoot => {
            const root = resolvedRoot || (await this.fileSystem.getRoots())[0];
            if (root) {
...
            }
        });
```

kittaakos replied:
Yeah, it makes sense, I just did not check your snippet thoroughly.

kittaakos commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -181,6 +183,11 @@ export class DugiteGit implements Git {
 
+    protected async getRootUri(): Promise<string> {
+        const root = await this.workspace.getRoot();
+        return root ? root : (await this.fileSystem.getRoots())[0].uri;
```
Then I'll change it. My Idea is to provide an empty array of repositories to the client when the workspace root is not yet set. Also, I will reject the `git clone` attempt.
